include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -8235,502 +7474,43 @@ FEFaceEvaluation<
-                        [this->active_fe_index][this->first_selected_component],
-                    grad,
-                    vector_selector);
-                  temp1[i + dofs_per_face + 2 * comp * dofs_per_face] =
-                    grad_weight * (temp1[i + 2 * comp * dofs_per_face] - grad);
-                }
-            }
-        }
-      else
-        {
-          AssertDimension(this->data->face_to_cell_index_nodal.size(1),
-                          dofs_per_face);
-          const unsigned int *index_array =
-            &this->data->face_to_cell_index_nodal(this->face_no, 0);
-          for (unsigned int i = 0; i < dofs_per_face; ++i)
-            for (unsigned int comp = 0; comp < n_components_; ++comp)
-              {
-                const unsigned int ind = index_array[i];
-                reader.process_dof_gather(
-                  indices,
-                  input_vector,
-                  ind + comp * static_dofs_per_component +
-                    this->dof_info->component_dof_indices_offset
-                      [this->active_fe_index][this->first_selected_component],
-                  temp1[i + comp * 2 * dofs_per_face],
-                  vector_selector);
-              }
-        }
-    }
-
-  // case 5: default vector access
-  else
+  static_assert(internal::has_begin<VectorType>::value &&
+                  (std::is_same<decltype(std::declval<VectorType>().begin()),
+                                double *>::value ||
+                   std::is_same<decltype(std::declval<VectorType>().begin()),
+                                float *>::value),
+                "This function requires a vector type with begin() function "
+                "evaluating to a pointer to basic number (float,double). "
+                "Use read_dof_values() followed by evaluate() instead.");
+
+  if (!internal::FEFaceEvaluationSelector<dim,
+                                          fe_degree,
+                                          n_q_points_1d,
+                                          n_components,
+                                          Number,
+                                          VectorizedArrayType>::
+        gather_evaluate(input_vector.begin(),
+                        *this->data,
+                        *this->dof_info,
+                        this->begin_values(),
+                        this->begin_gradients(),
+                        this->scratch_data,
+                        evaluate_values,
+                        evaluate_gradients,
+                        this->active_fe_index,
+                        this->first_selected_component,
+                        this->cell,
+                        this->face_no,
+                        this->subface_index,
+                        this->dof_access_index,
+                        this->face_orientation,
+                        this->mapping_data->descriptor[this->active_fe_index]
+                          .face_orientations))
```

Comment:
This spot from lines 7477-7508 is one of the large blocks where I outsourced to an implementation class.